### DISCUSSION TITLE
Share .env file with Vite and Express fullstack app

### DISCUSSION THREAD

**Discussion by OxyTJ** [vercel/turborepo#1799] [Help] [answered]
I'm trying to share a `.env` file across the monorepo. Initially I created a `dotenv.config.js` file inside a `config` package which contains the following code:

```
module.exports = require('dotenv').config({ path: `${__dirname}/.env` });
```

From the Express application, I can import it with the following:

```
require('config/dotenv.config');
```

While this works for backend apps, it does not work for the frontend Vite apps because accessing a path from the client code isn't permitted, which makes sense. I've tried adding the `.env` file to the monorepo root and add it as a `globalDependencies` to the `turbo.json`, however, this doesn't appear to do anything at all. For the Vite environment variables, I've prefixed them with `VITE_` as the documentation suggest and tried accessing them with `import.meta.env`, but they do not exist.

**Comment by OxyTJ** [comment, marked as answer]
I was able to get this to work with `vite.config.js` and using a relative path from each frontend app to the `.env` path, which is fine.

**Reply by shamashan** [reply]
Could you please provide more details on how you did this?. And is there any better solution now?

Thanks

**Reply by OxyTJ** [reply]
> Could you please provide more details on how you did this?. And is there any better solution now?
> 
> Thanks

It's been a few years, so I had to go back to the project I was working on at the time. This is how I had handled it:

```
import fs from 'fs';
import { defineConfig } from 'vite';

export default () => {
  const config = JSON.parse(
    fs.readFileSync(`${process.cwd()}/../../packages/configs/config.json`, {
      encoding: 'utf8',
    })
  );
  const { api } = config;

  process.env = {
    ...process.env,
    VITE_API_BASE_URL: `${api.baseUrl}:${api.port}`,
  };

  return defineConfig({
    server: {
      port: Number(api.port),
    },
  });
};
```

Coincidentally, this had been the last time I worked with Turborepo until just a couple of weeks ago and with this new project, I handle it a bit differently. Now, I store a template vite config file in the configs package. Example:

`packages/configs/env/dotenv.config.js`

```
import { fileURLToPath } from 'url';
import { dirname } from 'path';

import dotenv from 'dotenv';
import dotenvExpand from 'dotenv-expand';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const env = dotenv.config({ path: `${__dirname}/.env` });

export default dotenvExpand.expand(env);
```

`packages/configs/vite/vite.config.ts`

```
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import '../env/dotenv.config.js';

// only import the VITE-related environment variables
const viteEnv = Object.fromEntries(
  Object.entries(process.env)
    .filter(([key]) => key.startsWith('VITE_'))
    .map(([key, val]) => [`import.meta.env.${key}`, JSON.stringify(val)]),
);

// https://vite.dev/config/
export default defineConfig({
  plugins: [react()],
  define: viteEnv,
});
```

`packages/configs/package.json`

```
{
  "name": "@repo/configs",
  ...
  "exports": {
   ...
    "./vite/vite.config": "./vite/vite.config.js"
  }
}
```

`apps/web/vite.config.ts`

```
export { default } from '@repo/configs/vite/vite.config';
```
